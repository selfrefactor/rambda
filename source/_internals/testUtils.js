import combinate from 'combinate'
import {
  equals,
  filter,
  forEach,
  init,
  last,
  omit,
  switcher,
  type,
} from 'rambdax'

const omitOk = omit('ok')

const PENDING = 'PENDING'
const RESULTS_EQUAL = 'results are equal'
const RESULTS_MISMATCH = 'results are different'
const ERRORS_EQUAL = 'errors are equal'
const ERRORS_MESSAGE_MISMATCH = 'errors messages are different'
const ERRORS_TYPE_MISMATCH = 'errors types are different'
const SHOULD_THROW = 'Rambda should throw'
const SHOULD_NOT_THROW = 'Rambda should not throw'
const ALL_ERROR_LABELS = {
  RESULTS_MISMATCH,
  ERRORS_MESSAGE_MISMATCH,
  ERRORS_TYPE_MISMATCH,
  SHOULD_NOT_THROW,
  SHOULD_THROW,
}
const MISSING = 'Missing error handle in parseError'

function parseError(err){
  const typeError = switcher(err)
    .is(x => x instanceof TypeError, 'TypeError')
    .is(x => x instanceof SyntaxError, 'SyntaxError')
    .is(x => x instanceof RangeError, 'RangeError')
    .is(x => x instanceof Error, 'Error')
    .default(MISSING)

  if (typeError === MISSING){
    throw new Error('typeError === MISSING')
  }

  return {
    message : err.message,
    type    : typeError,
    ok      : true,
  }
}

function executeSync(
  fn, inputs, returnsFunctionFlag
){
  let result = PENDING
  let error = { ok : false }
  try {
    result = returnsFunctionFlag ?
      fn(...init(inputs))(last(inputs)) :
      fn(...inputs)
  } catch (e){
    error = parseError(e)
  }

  return {
    result,
    error,
  }
}

async function executeAsync(
  fn, inputs, returnsFunctionFlag
){
  let result = PENDING
  let error = { ok : false }
  try {
    result = returnsFunctionFlag ?
      await fn(...init(inputs))(last(inputs)) :
      await fn(...inputs)
  } catch (e){
    error = parseError(e)
  }

  return {
    result,
    error,
  }
}

export function profileMethod({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  returnsFunctionFlag = false,
  fn,
}){
  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)

    test(getTestTitle(...inputs), () => {
      const { result, error } = executeSync(
        fn, inputs, returnsFunctionFlag
      )

      expect({
        result,
        error : error.ok ? omitOk(error) : PENDING,
        inputs,
      }).toMatchSnapshot()
    })
  })
}

export function profileMethodAsync({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  returnsFunctionFlag = false,
  fn,
}){
  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)
    test(getTestTitle(...inputs), async () => {
      const { result, error } = await executeAsync(
        fn,
        inputs,
        returnsFunctionFlag
      )

      expect({
        result,
        error : error.ok ? omitOk(error) : PENDING,
        inputs,
      }).toMatchSnapshot()
    })
  })
}

export function compareToRamda(
  fn, fnRamda, returnsFunctionFlag
){
  return (...inputs) => {
    const { result, error } = executeSync(
      fn, inputs, returnsFunctionFlag
    )
    const { result: ramdaResult, error: ramdaError } = executeSync(
      fnRamda,
      inputs,
      returnsFunctionFlag
    )

    const toReturn = {
      result,
      ramdaResult,
      ramdaError : ramdaError.ok ? omitOk(ramdaError) : PENDING,
      error      : error.ok ? omitOk(error) : PENDING,
    }

    if (result !== PENDING){
      if (ramdaError.ok){
        return {
          ...toReturn,
          ok    : false,
          label : SHOULD_THROW,
        }
      }

      if (equals(result, ramdaResult)){
        return {
          ...toReturn,
          ok    : true,
          label : RESULTS_EQUAL,
        }
      }

      return {
        ...toReturn,
        ok    : false,
        label : RESULTS_MISMATCH,
      }
    }

    if (equals(error, ramdaError)){
      return {
        ...toReturn,
        ok    : true,
        label : ERRORS_EQUAL,
      }
    }

    if (ramdaError.ok){
      return {
        ...toReturn,
        ok    : false,
        label :
          ramdaError.type === error.type ?
            ERRORS_MESSAGE_MISMATCH :
            ERRORS_TYPE_MISMATCH,
      }
    }

    return {
      ...toReturn,
      ok    : false,
      label : SHOULD_NOT_THROW,
    }
  }
}

export const getTestTitle = (...inputs) => inputs.map(type).join(' | ')

export const compareCombinations = ({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  returnsFunctionFlag = false,
  setCounter = () => {},
  callback = x => {},
  fn,
  fnRamda,
}) => {
  const counter = {
    RESULTS_MISMATCH        : 0,
    SHOULD_THROW            : 0,
    SHOULD_NOT_THROW        : 0,
    ERRORS_TYPE_MISMATCH    : 0,
    ERRORS_MESSAGE_MISMATCH : 0,
    TOTAL_TESTS             : 0,
  }

  const increaseCounter = comparedResult => {
    if (comparedResult.ok) return
    let counterProp
    forEach((x, prop) => {
      if (x === comparedResult.label) counterProp = prop
    }, ALL_ERROR_LABELS)
    counter[ counterProp ]++
  }

  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)
  const compareOutputs = compareToRamda(
    fn, fnRamda, returnsFunctionFlag
  )

  afterAll(() => callback(counter))

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)

    test(getTestTitle(...inputs), () => {
      const compared = compareOutputs(...inputs)
      setCounter()
      if (!compared.ok){
        // if (compared.label === RESULTS_MISMATCH){
        increaseCounter(compared)
        expect({
          ...compared,
          inputs,
        }).toMatchSnapshot()
      }
      counter.TOTAL_TESTS++
    })
  })
}

const demoMap = new Map()

demoMap.set('a', 1)
demoMap.set('b', 2)

const demoWeakMap = new WeakMap()

demoWeakMap.set({ a : 0 }, 1)
demoWeakMap.set({ b : 0 }, 2)

const demoWeakSet = new WeakSet()
const foo = { a : 1 }
const bar = { b : 2 }

demoWeakSet.add(foo)
demoWeakSet.add(bar)

const demoBigInt = BigInt(9007199254740991)

const demoRegExp = new RegExp('ab+c', 'i')
const demoSymbol = Symbol('foo')

export const FALSY_VALUES = [ null, undefined, false, NaN, '' ]
export const EXTRA_BUILD_IN_OBJECTS = [
  Infinity,
  -Infinity,
  Symbol,
  new ArrayBuffer(8),
  demoRegExp,
  demoMap,
  demoWeakMap,
  demoWeakSet,
  demoBigInt,
  demoSymbol,
]
