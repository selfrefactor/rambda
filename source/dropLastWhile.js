export function dropLastWhile(predicate) {
  return list => {
    if (list.length === 0) {
      return list
    }

    const toReturn = []
    let counter = list.length

    while (counter) {
      const item = list[--counter]
      if (!predicate(item, counter)) {
        toReturn.push(item)
        break
      }
    }

    while (counter) {
      toReturn.push(list[--counter])
    }

    return toReturn.reverse()
  }
}
