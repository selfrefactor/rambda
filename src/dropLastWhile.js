import {_isArray} from './_internals/_isArray'

export function dropLastWhile(predicate, iterable) {
  if (arguments.length === 1) {
    return _iterable => dropLastWhile(predicate, _iterable)
  }
  if (iterable.length === 0) return iterable
  const isArray = _isArray(iterable)

  if (typeof predicate !== 'function') {
    throw new Error(`'predicate' is from wrong type ${typeof predicate}`)
  }
  if (!isArray && typeof iterable !== 'string') {
    throw new Error(`'iterable' is from wrong type ${typeof iterable}`)
  }

  let found = false
  const toReturn = []
  let counter = iterable.length

  while (counter > 0) {
    counter--
    if (!found && predicate(iterable[counter]) === false) {
      found = true
      toReturn.push(iterable[counter])
    } else if (found) {
      toReturn.push(iterable[counter])
    }
  }

  return isArray ? toReturn.reverse() : toReturn.reverse().join('')
}
