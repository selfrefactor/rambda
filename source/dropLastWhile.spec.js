import { dropLastWhile as dropLastWhileRamda } from 'ramda'

import { compareCombinations } from './_internals/testUtils.js'
import { dropLastWhile } from './dropLastWhile.js'

const list = [ 1, 2, 3, 4, 5 ]
const str = 'foobar'

test('with list', () => {
  const result = dropLastWhile(x => x >= 3, list)
  expect(result).toEqual([ 1, 2 ])
})

test('with string', () => {
  const result = dropLastWhile(x => x !== 'b')(str)
  expect(result).toBe('foob')
})

test('with empty list', () => {
  expect(dropLastWhile(() => true, [])).toEqual([])
  expect(dropLastWhile(() => false, [])).toEqual([])
})

const possiblePredicates = [
  x => x > 2,
  x => x < 2,
  x => x < -2,
  x => x > 10,
  '',
  [],
  [ 1 ],
]

const possibleIterables = [
  list,
  [ {}, '1', 2 ],
  str,
  `${ str }${ str }`,
  /foo/g,
  Promise.resolve('foo'),
  2,
]

describe('brute force', () => {
  compareCombinations({
    fn          : dropLastWhile,
    fnRamda     : dropLastWhileRamda,
    firstInput  : possiblePredicates,
    secondInput : possibleIterables,
    callback    : errorsCounters => {
      expect(errorsCounters).toMatchInlineSnapshot(`
        {
          "ERRORS_MESSAGE_MISMATCH": 0,
          "ERRORS_TYPE_MISMATCH": 12,
          "RESULTS_MISMATCH": 0,
          "SHOULD_NOT_THROW": 21,
          "SHOULD_THROW": 0,
          "TOTAL_TESTS": 49,
        }
      `)
    },
  })
})
