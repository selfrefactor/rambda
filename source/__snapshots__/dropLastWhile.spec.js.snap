// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`brute force Array | Array 1`] = `
{
  "error": {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [],
    [
      1,
      2,
      3,
      4,
      5,
    ],
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array | Array 2`] = `
{
  "error": {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [
      1,
    ],
    [
      1,
      2,
      3,
      4,
      5,
    ],
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array | Array 3`] = `
{
  "error": {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [],
    [
      {},
      "1",
      2,
    ],
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array | Array 4`] = `
{
  "error": {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [
      1,
    ],
    [
      {},
      "1",
      2,
    ],
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array | Number 1`] = `
{
  "error": {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [],
    2,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Array | Number 2`] = `
{
  "error": {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [
      1,
    ],
    2,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Array | Promise 1`] = `
{
  "error": {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [],
    Promise {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Array | Promise 2`] = `
{
  "error": {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [
      1,
    ],
    Promise {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Array | RegExp 1`] = `
{
  "error": {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [],
    /foo/g,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Array | RegExp 2`] = `
{
  "error": {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [
      1,
    ],
    /foo/g,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Array | String 1`] = `
{
  "error": {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [],
    "foobar",
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array | String 2`] = `
{
  "error": {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [
      1,
    ],
    "foobar",
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array | String 3`] = `
{
  "error": {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [],
    "foobarfoobar",
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array | String 4`] = `
{
  "error": {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [
      1,
    ],
    "foobarfoobar",
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Function | Number 1`] = `
{
  "error": {
    "message": "'iterable' is from wrong type number",
    "type": "Error",
  },
  "inputs": [
    [Function],
    2,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Function | Number 2`] = `
{
  "error": {
    "message": "'iterable' is from wrong type number",
    "type": "Error",
  },
  "inputs": [
    [Function],
    2,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Function | Number 3`] = `
{
  "error": {
    "message": "'iterable' is from wrong type number",
    "type": "Error",
  },
  "inputs": [
    [Function],
    2,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Function | Number 4`] = `
{
  "error": {
    "message": "'iterable' is from wrong type number",
    "type": "Error",
  },
  "inputs": [
    [Function],
    2,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Function | Promise 1`] = `
{
  "error": {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [Function],
    Promise {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Function | Promise 2`] = `
{
  "error": {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [Function],
    Promise {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Function | Promise 3`] = `
{
  "error": {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [Function],
    Promise {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Function | Promise 4`] = `
{
  "error": {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [Function],
    Promise {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Function | RegExp 1`] = `
{
  "error": {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [Function],
    /foo/g,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Function | RegExp 2`] = `
{
  "error": {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [Function],
    /foo/g,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Function | RegExp 3`] = `
{
  "error": {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [Function],
    /foo/g,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Function | RegExp 4`] = `
{
  "error": {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": [
    [Function],
    /foo/g,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force String | Array 1`] = `
{
  "error": {
    "message": "'predicate' is from wrong type string",
    "type": "Error",
  },
  "inputs": [
    "",
    [
      1,
      2,
      3,
      4,
      5,
    ],
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force String | Array 2`] = `
{
  "error": {
    "message": "'predicate' is from wrong type string",
    "type": "Error",
  },
  "inputs": [
    "",
    [
      {},
      "1",
      2,
    ],
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force String | Number 1`] = `
{
  "error": {
    "message": "'predicate' is from wrong type string",
    "type": "Error",
  },
  "inputs": [
    "",
    2,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force String | Promise 1`] = `
{
  "error": {
    "message": "'predicate' is from wrong type string",
    "type": "Error",
  },
  "inputs": [
    "",
    Promise {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force String | RegExp 1`] = `
{
  "error": {
    "message": "'predicate' is from wrong type string",
    "type": "Error",
  },
  "inputs": [
    "",
    /foo/g,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force String | String 1`] = `
{
  "error": {
    "message": "'predicate' is from wrong type string",
    "type": "Error",
  },
  "inputs": [
    "",
    "foobar",
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force String | String 2`] = `
{
  "error": {
    "message": "'predicate' is from wrong type string",
    "type": "Error",
  },
  "inputs": [
    "",
    "foobarfoobar",
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;
